Compile fails on ESP8266

Hi,

I tried to compile it with nodemcu but it failed,

```
Build options changed, rebuilding all
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp: In constructor 'MedianFilter::MedianFilter(byte, int)':
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: error: no matching function for call to 'max(const byte&, int)'
  medFilterWin = max(size, 3);   // number of samples in sliding median filter window - usually odd #

                            ^
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: note: candidates are:
In file included from /home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/algorithm:62:0,
                 from /home/user/.arduino15/packages/esp8266/hardware/esp8266/2.4.1/cores/esp8266/Arduino.h:254,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.h:36,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:40:
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algo.h:4236:5: note: template<class _Tp, class _Compare> _Tp std::max(std::initializer_list<_Tp>, _Compare)
     max(initializer_list<_Tp> __l, _Compare __comp)
     ^
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algo.h:4236:5: note:   template argument deduction/substitution failed:
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: note:   mismatched types 'std::initializer_list<_Tp>' and 'unsigned char'
  medFilterWin = max(size, 3);   // number of samples in sliding median filter window - usually odd #

                            ^
In file included from /home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/algorithm:62:0,
                 from /home/user/.arduino15/packages/esp8266/hardware/esp8266/2.4.1/cores/esp8266/Arduino.h:254,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.h:36,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:40:
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algo.h:4231:5: note: template<class _Tp> _Tp std::max(std::initializer_list<_Tp>)
     max(initializer_list<_Tp> __l)
     ^
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algo.h:4231:5: note:   template argument deduction/substitution failed:
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: note:   mismatched types 'std::initializer_list<_Tp>' and 'unsigned char'
  medFilterWin = max(size, 3);   // number of samples in sliding median filter window - usually odd #

                            ^
In file included from /home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/algorithm:61:0,
                 from /home/user/.arduino15/packages/esp8266/hardware/esp8266/2.4.1/cores/esp8266/Arduino.h:254,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.h:36,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:40:
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algobase.h:260:5: note: template<class _Tp, class _Compare> const _Tp& std::max(const _Tp&, const _Tp&, _Compare)
     max(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algobase.h:260:5: note:   template argument deduction/substitution failed:
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: note:   deduced conflicting types for parameter 'const _Tp' ('unsigned char' and 'int')
  medFilterWin = max(size, 3);   // number of samples in sliding median filter window - usually odd #

                            ^
In file included from /home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/algorithm:61:0,
                 from /home/user/.arduino15/packages/esp8266/hardware/esp8266/2.4.1/cores/esp8266/Arduino.h:254,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.h:36,
                 from /home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:40:
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algobase.h:216:5: note: template<class _Tp> const _Tp& std::max(const _Tp&, const _Tp&)
     max(const _Tp& __a, const _Tp& __b)
     ^
/home/user/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/1.20.0-26-gb404fb9-2/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_algobase.h:216:5: note:   template argument deduction/substitution failed:
/home/user/Arduino/libraries/SharpDistSensor/MedianFilter.cpp:47:28: note:   deduced conflicting types for parameter 'const _Tp' ('unsigned char' and 'int')
  medFilterWin = max(size, 3);   // number of samples in sliding median filter window - usually odd #

                            ^
exit status 1
Error compiling for board NodeMCU 1.0 (ESP-12E Module).
 
```

--- a/MedianFilter.cpp
+++ b/MedianFilter.cpp
@@ -1,180 +1,207 @@
 /*
-	MedianFilter.cpp - Median Filter for the Arduino platform.
-	Copyright (c) 2013 Phillip Schmidt.  All right reserved.
+   MedianFilter.cpp - Median Filter for the Arduino platform.
+   Copyright (c) 2013 Phillip Schmidt.  All right reserved.
 
-	This library is free software; you can redistribute it and/or
-	modify it under the terms of the GNU Lesser General Public
-	License as published by the Free Software Foundation; either
-	version 2.1 of the License, or (at your option) any later version.
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
 
-	This library is distributed in the hope that it will be useful,
-	but WITHOUT ANY WARRANTY; without even the implied warranty of
-	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-	Lesser General Public License for more details.
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
 
-	You should have received a copy of the GNU Lesser General Public
-	License along with this library; if not, write to the Free Software
-	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
  */
 
-/* 
- 	A median filter object is created by by passing the desired filter window size on object creation.
-	The window size should be an odd number between 3 and 255.  
-  
-	New data is added to the median filter by passing the data through the in() function.  The new medial value is returned.
-	The new data will over-write the oldest data point, then be shifted in the array to place it in the correct location. 
-  
-	The current median value is returned by the out() function for situations where the result is desired without passing in new data.
- 
-	!!! All data must be type INT.  !!!
-	
-	Window Size / avg processing time [us]
-	5  / 22
-	7  / 30
-	9  / 40
-	11 / 49
-	21 / 99
-	
+/*
+    A median filter object is created by by passing the desired filter window size on object creation.
+   The window size should be an odd number between 3 and 255.
+
+   New data is added to the median filter by passing the data through the in() function.  The new medial value is returned.
+   The new data will over-write the oldest data point, then be shifted in the array to place it in the correct location.
+
+   The current median value is returned by the out() function for situations where the result is desired without passing in new data.
+
+   !!! All data must be type INT.  !!!
+
+   Window Size / avg processing time [us]
+   5  / 22
+   7  / 30
+   9  / 40
+   11 / 49
+   21 / 99
+
 */
 
 #include "MedianFilter.h"
 
-	
 
-MedianFilter::MedianFilter(const byte size, const int seed)
+MedianFilter::MedianFilter(int size, int seed)
 {
+   medFilterWin    = constrain(size, 3, 255); // number of samples in sliding median filter window - usually odd #
+   medDataPointer  = size >> 1;           // mid point of window
+   data            = (int*)     calloc (size, sizeof(int));     // array for data
+   sizeMap         = (uint8_t*) calloc (size, sizeof(uint8_t)); // array for locations of data in sorted list
+   locationMap     = (uint8_t*) calloc (size, sizeof(uint8_t)); // array for locations of history data in map list
+   oldestDataPoint = medDataPointer;      // oldest data point location in data array
+   totalSum        = size * seed;         // total of all values
 
-	medFilterWin = max(size, 3);			// number of samples in sliding median filter window - usually odd #
-	medDataPointer = size >> 1;				// mid point of window
-	data = (int*) calloc (size, sizeof(int));			// array for data
-	sizeMap = (byte*) calloc (size, sizeof(byte));		// array for locations of data in sorted list
-	locationMap = (byte*) calloc (size, sizeof(byte));	// array for locations of history data in map list
-	oldestDataPoint = medDataPointer;					// oldest data point location in data array
-	
-	for(byte i = 0; i < medFilterWin; i++) // initialize the arrays
-	{
-		sizeMap[i] = i;			// start map with straight run
-		locationMap[i] = i;		// start map with straight run
-		data[i] = seed;			// populate with seed value
-	}
-	
+   for(uint8_t i = 0; i < medFilterWin; i++) // initialize the arrays
+   {
+      sizeMap[i]     = i;      // start map with straight run
+      locationMap[i] = i;      // start map with straight run
+      data[i]        = seed;   // populate with seed value
+   }
 }
 
-int MedianFilter::in(int value)
-{  
-	// sort sizeMap
-	// small vaues on the left (-)
-	// larger values on the right (+)
 
-	boolean dataMoved = false;
-	const byte rightEdge = medFilterWin - 1;  // adjusted for zero indexed array
-	
-	data[oldestDataPoint] = value;  // store new data in location of oldest data in ring buffer
-	
-	// SORT LEFT (-) <======(n) (+)
-	if(locationMap[oldestDataPoint] > 0){ // don't check left neighbours if at the extreme left
-		
-		for(byte i = locationMap[oldestDataPoint]; i > 0; i--){	//index through left adjacent data
-			
-			byte n = i - 1;	// neighbour location
-			
-			if(data[oldestDataPoint] < data[sizeMap[n]]){ // find insertion point, move old data into position
+int MedianFilter::in(const int & value)
+{
+   // sort sizeMap
+   // small vaues on the left (-)
+   // larger values on the right (+)
 
-				sizeMap[i] = sizeMap[n];	// move existing data right so the new data can go left
-				locationMap[sizeMap[n]]++;
-				
-				sizeMap[n] = oldestDataPoint; // assign new data to neighbor position
-				locationMap[oldestDataPoint]--;
-				
-				dataMoved = true;
-				
-			}
-			else {
+   boolean dataMoved = false;
+   const uint8_t rightEdge = medFilterWin - 1;  // adjusted for zero indexed array
 
-				break; // stop checking once a smaller value is found on the left 
-				
-			}
-		}
-	}
-	
-	// SORT RIGHT (-) (n)======> (+)
-	if(!dataMoved && locationMap[oldestDataPoint] < rightEdge){ // don't check right if at right border, or the data has already moved
+   totalSum += value - data[oldestDataPoint];  // add new value and remove oldest value
 
-		for(int i = locationMap[oldestDataPoint]; i < rightEdge; i++){	//index through left adjacent data
+   data[oldestDataPoint] = value;  // store new data in location of oldest data in ring buffer
 
-			int n = i + 1;	// neighbour location
-			
-			if(data[oldestDataPoint] > data[sizeMap[n]]){ // find insertion point, move old data into position
+   // SORT LEFT (-) <======(n) (+)
+   if(locationMap[oldestDataPoint] > 0) // don't check left neighbours if at the extreme left
+   {
+      for(uint8_t i = locationMap[oldestDataPoint]; i > 0; i--)   //index through left adjacent data
+      {
+         uint8_t n = i - 1;   // neighbour location
 
-				sizeMap[i] = sizeMap[n];	// move existing data left so the new data can go right
-				locationMap[sizeMap[n]]--;
-				
-				sizeMap[n] = oldestDataPoint; // assign new data to neighbor position
-				locationMap[oldestDataPoint]++;
+         if(data[oldestDataPoint] < data[sizeMap[n]]) // find insertion point, move old data into position
+         {
+            sizeMap[i] = sizeMap[n];   // move existing data right so the new data can go left
+            locationMap[sizeMap[n]]++;
 
-			}
-			else {
+            sizeMap[n] = oldestDataPoint; // assign new data to neighbor position
+            locationMap[oldestDataPoint]--;
 
-				break; // stop checking once a smaller value is found on the right 
-				
-			}	
-		}
-	}	
-	
-	oldestDataPoint++; 		// increment and wrap
-	if(oldestDataPoint == medFilterWin) 
-		oldestDataPoint = 0;
-	
-	return data[sizeMap[medDataPointer]];
+            dataMoved = true;
+         }
+         else
+         {
+            break; // stop checking once a smaller value is found on the left
+         }
+      }
+   }
+
+   // SORT RIGHT (-) (n)======> (+)
+   if(!dataMoved && locationMap[oldestDataPoint] < rightEdge) // don't check right if at right border, or the data has already moved
+   {
+      for(int i = locationMap[oldestDataPoint]; i < rightEdge; i++)   //index through left adjacent data
+      {
+         int n = i + 1;   // neighbour location
+
+         if(data[oldestDataPoint] > data[sizeMap[n]]) // find insertion point, move old data into position
+         {
+            sizeMap[i] = sizeMap[n];   // move existing data left so the new data can go right
+            locationMap[sizeMap[n]]--;
+
+            sizeMap[n] = oldestDataPoint; // assign new data to neighbor position
+            locationMap[oldestDataPoint]++;
+         }
+         else
+         {
+            break; // stop checking once a smaller value is found on the right
+         }
+      }
+   }
+   oldestDataPoint++;       // increment and wrap
+   if(oldestDataPoint == medFilterWin) oldestDataPoint = 0;
+
+   return data[sizeMap[medDataPointer]];
 }
 
 
 int MedianFilter::out() // return the value of the median data sample
 {
-	return  data[sizeMap[medDataPointer]];     
+   return  data[sizeMap[medDataPointer]];
 }
 
+
+int MedianFilter::getMin()
+{
+   return data[sizeMap[ 0 ]];
+}
+
+
+int MedianFilter::getMax()
+{
+   return data[sizeMap[ medFilterWin - 1 ]];
+}
+
+
+int MedianFilter::getMean()
+{
+   return totalSum / medFilterWin;
+}
+
+
+int MedianFilter::getStDev()  // Arduino run time [us]: filterSize * 2 + 131
+{
+   int32_t diffSquareSum = 0;
+   int mean = getMean();
+
+   for( int i = 0; i < medFilterWin; i++ )
+   {
+      int diff = data[i] - mean;
+      diffSquareSum += diff * diff;
+   }
+
+   return int( sqrtf( float(diffSquareSum) / float(medFilterWin - 1) ) + 0.5f );
+}
 
 
 // *** debug fuctions ***
 /*
 void MedianFilter::printData() // display sorting data for debugging
 {
-	for(int i=0; i<medFilterWin; i++)
-	{	
-		Serial.print(data[i]);
-		Serial.print("\t");
-	}
-	Serial.println("Data in ring buffer");
+   for(int i=0; i<medFilterWin; i++)
+   {
+      Serial.print(data[i]);
+      Serial.print("\t");
+   }
+   Serial.println("Data in ring buffer");
 }
 
 void MedianFilter::printSizeMap()
-{ 
-	for(int i=0; i<medFilterWin; i++)
-	{
-		Serial.print(sizeMap[i]);
-		Serial.print("\t");
-	}
-	Serial.println("Size Map, data sorted by size");
+{
+   for(int i=0; i<medFilterWin; i++)
+   {
+      Serial.print(sizeMap[i]);
+      Serial.print("\t");
+   }
+   Serial.println("Size Map, data sorted by size");
 }
 
 void MedianFilter::printLocationMap()
-{	
-	for(int i=0; i<medFilterWin; i++)
-	{
-		Serial.print(locationMap[i]);
-		Serial.print("\t");
-	}
-	Serial.println("Location Map, size data sorted by age");
-}  
+{
+   for(int i=0; i<medFilterWin; i++)
+   {
+      Serial.print(locationMap[i]);
+      Serial.print("\t");
+   }
+   Serial.println("Location Map, size data sorted by age");
+}
 
 void MedianFilter::printSortedData() // display data for debugging
 {
-	for(int i=0; i<medFilterWin; i++)
-	{
-		Serial.print(data[sizeMap[i]]);
-		Serial.print("\t");
-	}
-	Serial.println("Data sorted by size");
+   for(int i=0; i<medFilterWin; i++)
+   {
+      Serial.print(data[sizeMap[i]]);
+      Serial.print("\t");
+   }
+   Serial.println("Data sorted by size");
 }
 */

--- a/MedianFilter.h
+++ b/MedianFilter.h
@@ -17,43 +17,51 @@
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
 
-/* 
-	A median filter object is created by by passing the desired filter window size on object creation.
-	The window size should be an odd number between 3 and 255.  
-  
-	New data is added to the median filter by passing the data through the in() function.  The new medial value is returned.
-	The new data will over-write the oldest data point, then be shifted in the array to place it in the correct location. 
-  
-	The current median value is returned by the out() function for situations where the result is desired without passing in new data.
- 
-	!!! All data must be type INT.  !!!
+/*
+   A median filter object is created by by passing the desired filter window size on object creation.
+   The window size should be an odd number between 3 and 255.
+
+   New data is added to the median filter by passing the data through the in() function.  The new medial value is returned.
+   The new data will over-write the oldest data point, then be shifted in the array to place it in the correct location.
+
+   The current median value is returned by the out() function for situations where the result is desired without passing in new data.
+
+   !!! All data must be type INT.  !!!
  */
 
 #ifndef MedianFilter_h
 
-	#define MedianFilter_h
+   #define MedianFilter_h
 
-	#include "Arduino.h"
+   #include "Arduino.h"
 
-	class MedianFilter
-	{
-		public:
-			MedianFilter(byte size, int seed);
-			int in(int value);
-			int out(); 
-			
-			//void printData();		// used for debugging
-			//void printSizeMap();
-			//void printLocationMap();
-			//void printSortedData();
-			
-		private:
-			byte medFilterWin;		// number of samples in sliding median filter window - usually odd #
-			byte medDataPointer;	// mid point of window
-			int  * data;			// array pointer for data sorted by age in ring buffer
-			byte * sizeMap;			// array pointer for locations data in sorted by size
-			byte * locationMap;		// array pointer for data locations in history map
-			byte oldestDataPoint;	// oldest data point location in ring buffer
-	};
+   class MedianFilter
+   {
+      public:
+         MedianFilter(int size, int seed);
+         int in(const int & value);
+         int out();
+
+         int getMin();
+         int getMax();
+         int getMean();
+         int getStDev();
+
+         /*
+         void printData();		// used for debugging
+         void printSizeMap();
+         void printLocationMap();
+         void printSortedData();
+         */
+
+      private:
+         uint8_t medFilterWin;      // number of samples in sliding median filter window - usually odd #
+         uint8_t medDataPointer;	   // mid point of window
+         int     * data;			   // array pointer for data sorted by age in ring buffer
+         uint8_t * sizeMap;			// array pointer for locations data in sorted by size
+         uint8_t * locationMap;		// array pointer for data locations in history map
+         uint8_t oldestDataPoint;	// oldest data point location in ring buffer
+         int32_t totalSum;
+   };
 
 #endif

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 # SharpDistSensor
 A library for the Arduino IDE that helps interface with Sharp analog distance sensors.  
 
-Version 1.4.0  
+Version 1.4.1  
 [![Build Status](https://travis-ci.org/DrGFreeman/SharpDistSensor.svg?branch=master)](https://travis-ci.org/DrGFreeman/SharpDistSensor)  
 By Julien de la Bruère-Terreault (drgfreeman@tuta.io)
 
@@ -84,6 +84,7 @@ Model | valMin | valMax
 This library has been designed so that it is easy to add sensor models. Contributions are therefore welcome. Adding models to the library can be done by either submitting a pull request or providing me the proposed fit function and associated calibration data by email so I can add it myself. Thank you for contributing!
 
 ## Version history
+* 1.4.1 (2018-08-07): Updated MedianFilter to latest version from [daPhoosa/MedianFilter](https://github.com/daPhoosa/MedianFilter).
 * 1.4.0 (2018-05-21): Added GP2Y0A51SK0F_5V_DS model.
 * 1.3.0 (2018-05-20): Added SharpDistSensorArray example.
 * 1.2.0 (2017-05-10): Added GP2Y0A710K0F_5V_DS model.

--- a/library.properties
+++ b/library.properties
@@ -1,5 +1,5 @@
 name=SharpDistSensor
-version=1.4.0
+version=1.4.1
 author=Julien de la Bruere-Terreault, drgfreeman@tuta.io
 maintainer=Julien de la Bruere-Terreault, drgfreeman@tuta.io
 sentence=Sharp analog distance sensor library

